src/Oscoin/Protocol.hs

```diff
@@ -0,0 +1,275 @@
+    -- enforced.
+    extendsTip <- do
+        currentTip <- BlockStore.getTip bsPublicAPI
+        pure $ blockHash currentTip == Block.parentHash incomingBlock
+
+    (mgr', evts) <-
+       if extendsTip
+          then do
+              -- Get the last (cached) 'mutableChainDepth' blocks and performs
+              -- full block validation.
+              let depth = fromIntegral $ mutableChainDepth (protoConfig mgr)
+              ancestors <- BlockStore.getBlocksByDepth bsPublicAPI depth
+              case protoValidateFull mgr (toNewestFirst ancestors) incomingBlock of
+                  Left validationError ->
+                      let evt = BlockValidationFailedEvent (blockHash incomingBlock) validationError
+                      in pure (mgr, [evt])
+                  Right () -> do
+                      -- Step 3: store the fully-validated block.
+                      BlockStore.insertBlock bsPrivateAPI incomingBlock
+                      pure (mgr, mempty)
+
+          -- NOTE(adn) This is a side effect of the fact that storing
+          -- the genesis as an orphan would insert it into the orphanage with
+          -- its parent hash, which means that we won't be able to retrieve
+          -- a potentiol candidate branching off from genesis in our 'selectBestChain'
+          -- function (cfr. QuickCheck seed 554786). To avoid that, we avoid
+          -- inserting the genesis block into the orphanage in the first place;
+          -- After all, the genesis block should never be considered at orphan.
+          else map (,mempty) . pure $
+              if not (isGenesisBlock incomingBlock)
+                  then mgr { protoOrphanage = O.insertOrphan incomingBlock (protoOrphanage mgr) }
+                  else mgr
+
+    -- Step 4: Chain selection.
+    (,evts) <$> selectBestChain mgr'
+  where
+      bsPublicAPI :: BlockStoreReader c tx s m
+      bsPublicAPI = fst . protoFullBlockStore $ mgr
+
+      bsPrivateAPI :: BlockStoreWriter c tx s m
+      bsPrivateAPI = snd . protoFullBlockStore $ mgr
+
+{------------------------------------------------------------------------------
+  (Best) Chain selection
+------------------------------------------------------------------------------}
+
+-- | Performs best chain selection and returns the new 'Protocol' data
+-- structure.
+selectBestChain
+    :: forall m s c tx.
+       ( Monad m
+       , Ord s
+       , Ord (BlockHash c)
+       )
+    => Protocol c tx s m
+    -> m (Protocol c tx s m)
+selectBestChain mgr = do
+    -- NOTE(adn) Grab the \"mutable\" part of the chain and
+    -- compare each root hash to see if there is any fork originating from that.
+    let depth = fromIntegral $ mutableChainDepth (protoConfig mgr)
+    mutableBlockHashes <-
+        map blockHash . Chrono.reverse <$> BlockStore.getBlocksByDepth bsPublicAPI depth
+
+    case O.selectBestChain (toOldestFirst mutableBlockHashes) orphanage of
+        Just (parentHash, bestChain) | isValid bestChain -> do
```

Comment:
Do we have to check `isValid` here? Isn't that really expensive? I'd assume that chains returned from the orphanage are always valid.

Reply:
There's also the issue that if the best chain is invalid, then perhaps there was another chain that scored lesser, but is actually valid.

Reply:
> Do we have to check isValid here? Isn't that really expensive?

Uhm, but wasn't one of the objective of this whole refactoring split validation in two halves where the heavy part (i.e. this one) would be done only at this stage?

>  I'd assume that chains returned from the orphanage are always valid.

When we insert a block in the orphanage we check only "phase-1" validation, so it's actually possible that the overall best chain is not valid when we try to hook in into the main chain, no?

> There's also the issue that if the best chain is invalid, then perhaps there was another chain that scored lesser, but is actually valid.

Ah, fair point. Yes, this was something I didn't have to worry about in the first iteration of fork selection, but this is now missing from this patch.

Reply:
I am also thinking about something else -- in particular, we are calling `validateBlockchain` without taking a look at the prefix this new chain will insert onto. In other terms, say you have found a fork, and you need to validate such fork. shouldn't we also make sure that the fork _plus_ a bunch of blocks in the main chain also validate? Otherwise it might be possible that the fork is _locally_ valid, but not "globally", if you see what I mean?

Did I lose the plot? 🤔

Reply:
Ok it seems like this is a bigger conversation - can we create an issue and link to this comment? I don't want to hold up the PR.

Reply:
Yes, sir 👍

Reply:
https://github.com/oscoin/oscoin/issues/420